File: Java/Ema/Src/main/java/interface/com/thomsonreuters/ema/access/RefreshMsg.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright Thomson Reuters 2015. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.access;

import java.nio.ByteBuffer;

/**
 * RefreshMsg conveys item image, state, permission and group information.
 * 
 * <p>RefreshMsg is sent when item data needs to be synchronized.
 * This happens as a response to received ReqMsg or when upstream source requires it.</p>
 * <p>RefreshMsg sent as a response to ReqMsg is called a solicited refresh,
 * while an unsolicited refresh is sent when upstream source requires
 * synchronization of downstream consumers.</p>
 * 
 * @see Data
 * @see Msg
 */
public interface RefreshMsg extends Msg
{
	/**
	 * Indicates presence of Qos.
	 * Qos is an optional member of RefreshMsg.
	 * 
	 * @return true if Qos is set; false otherwise
	 */
	public boolean hasQos();

	/**
	 * Indicates presence of SeqNum.
	 * Sequence number (SeqNum) is an optional member of RefreshMsg.
	 * 
	 * @return true if sequence number is set; false otherwise
	 */
	public boolean hasSeqNum();

	/**
	 * Indicates presence of PartNum.
	 * Part number is an optional member of RefreshMsg.
	 * 
	 * @return true if part number is set; false otherwise
	 */
	public boolean hasPartNum();

	/**
	 * Indicates presence of PermissionData.
	 * Permission data is optional member of RefreshMsg.
	 * 
	 * @return true if permission data is set; false otherwise
	 */
	public boolean hasPermissionData();

	/**
	 * Indicates presence of PublisherId.
	 * Publisher id is an optional member of RefreshMsg.
	 * 
	 * @return true if publisher id is set; false otherwise
	 */
	public boolean hasPublisherId();

	/**
	 * Indicates presence of the ServiceName within the MsgKey.
	 * Service name is an optional member of RefreshMsg.
	 * 
	 * @return true if service name is set; false otherwise
	 */
	public boolean hasServiceName();

	/**
	 * Returns State ({@link com.thomsonreuters.ema.access.OmmState}).
	 * 
	 * @return state of item (e.g. Open / Ok)
	 */
	public OmmState state();

	/**
	 * Returns Qos ({@link com.thomsonreuters.ema.access.OmmQos}).
	 * <br>Calling this method must be preceded by a call to {@link #hasQos()}.
	 * 
	 * @throws OmmInvalidUsageException if {@link #hasQos()} returns false
	 * 
	 * @return Qos of item
	 */
	public OmmQos qos();

	/**
	 * Returns SeqNum.
	 * <br>Calling this method must be preceded by a call to {@link #hasSeqNum()}.
	 * 
	 * @throws OmmInvalidUsageException if {@link #hasSeqNum()} returns false
	 * 
	 * @return sequence number
	 */
	public long seqNum();

	/**
	 * Returns PartNum.
	 * <br>Calling this method must be preceded by a call to {@link #hasPartNum()}
	 * 
	 * @throws OmmInvalidUsageException if {@link #hasPartNum()} returns false
	 * 
	 * @return part number
	 */
	public int partNum();

	/**
	 * Returns ItemGroup.
	 * 
	 * @return ByteBuffer containing item group information
	 */
	public ByteBuffer itemGroup();

	/**
	 * Returns PermissionData.
 	 * <br>Calling this method must be preceded by a call to {@link #hasPermissionData()}
	 * 
	 * @throws OmmInvalidUsageException if {@link #hasPermissionData()} returns false
	 * 
	 * @return ByteBuffer containing permission data
	 */
	public ByteBuffer permissionData();

	/**
	 * Returns PublisherIdUserId.
	 * 
	 * @throws OmmInvalidUsageException if {@link #hasPublisherId()} returns false
	 * 
	 * @return publisher's user Id
	 */
	public long publisherIdUserId();

	/**
	 * Returns PublisherIdUserAddress.
  	 * <br>Calling this method must be preceded by a call to {@link #hasPublisherId()}
	 * 
	 * @throws OmmInvalidUsageException if {@link #hasPublisherId()} returns false
	 * 
	 * @return publisher's user address
	 */
	public long publisherIdUserAddress();

	/**
	 * Returns Solicited.
	 * 
	 * @return true if this is solicited refresh; false otherwise
	 */
	public boolean solicited();

	/**
	 * Returns DoNotCache.
	 * 
	 * @return true if this refresh must not be cached; false otherwise
	 */
	public boolean doNotCache();

	/**
	 * Returns Complete.
	 * 
	 * @return true if this is the last part of the multi part refresh message
	 */
	public boolean complete();

	/**
	 * Returns ClearCache.
	 * 
	 * @return true if cache needs to be cleared before applying this refresh;
	 *         false otherwise
	 */
	public boolean clearCache();

	/**
	 * Returns PrivateStream.
	 * 
	 * @return true if this is private stream item
	 */
	public boolean privateStream();

	/**
	 * Returns the ServiceName within the MsgKey.
	 * <br>Calling this method must be preceded by a call to {@link #hasServiceName()}
	 * 
	 * @throws OmmInvalidUsageException if {@link #hasServiceName()} returns false
	 * 
	 * @return String containing service name
	 */
	public String serviceName();

	/**
	 * Clears the RefreshMsg.
	 * Invoking clear() method clears all the values and resets all the defaults.
	 * 
	 * @return reference to this object
	 */
	public RefreshMsg clear();

	/**
	 * Specifies StreamId.
	 * 
	 * @param streamId stream id
	 * @return reference to this object
	 */
	public RefreshMsg streamId(int streamId);

	/**
	 * Specifies Domain Type.
	 * 
	 * @throws OmmUnsupportedDomainTypeException if domainType is greater than 255
	 * 
	 * @param domainType specifies RDM Message Model Type
	 *        (default value is {@link com.thomsonreuters.ema.rdm.EmaRdm#MMT_MARKET_PRICE})
	 * @return reference to this object
	 */
	public RefreshMsg domainType(int domainType);

	/**
	 * Specifies Name.
	 * 
	 * @param name an String object containing item name
	 * @return reference to this object
	 */
	public RefreshMsg name(String name);

	/**
	 * Specifies NameType.
	 * 
	 * @param nameType specifies RDM Instrument NameType 
	 *        (default value is {@link com.thomsonreuters.ema.rdm.EmaRdm#INSTRUMENT_NAME_RIC})
	 * @return reference to this object
	 */
	public RefreshMsg nameType(int nameType);

	/**
	 * Specifies ServiceName.
	 * 
	 * @throws OmmInvalidUsageException if service id is already set
	 * 
	 * @param serviceName an String object containing service name
	 * @return reference to this object
	 */
	public RefreshMsg serviceName(String serviceName);

	/**
	 * Specifies ServiceId.
	 * 
	 * @throws OmmInvalidUsageException if service name is already set
	 * 
	 * @param serviceId service identifier
	 * @return reference to this object
	 */
	public RefreshMsg serviceId(int serviceId);

	/**
	 * Specifies Id.
	 * 
	 * @param id specifies Id
	 * @return reference to this object
	 */
	public RefreshMsg id(int id);

	/**
	 * Specifies Filter.
	 * 
	 * @param filter specifies Filter
	 * @return reference to this object
	 */
	public RefreshMsg filter(long filter);

	/**
	 * Specifies Qos.
	 * 
	 * @param timeliness specifies Qos Timeliness
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmQos.Timeliness#REALTIME})
	 * @param rate specifies Qos rate
	 *        (default value is	{@link com.thomsonreuters.ema.access.OmmQos.Rate#TICK_BY_TICK})
	 * @return reference to this object
	 */
	public RefreshMsg qos(int timeliness, int rate);

	/**
	 * Specifies State.
	 * 
	 * @param streamState conveys item stream state value
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.StreamState#OPEN})
	 * @param dataState conveys item data state value
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.DataState#SUSPECT})
	 * @return reference to this object
	 */
	public RefreshMsg state(int streamState, int dataState);
	
	/**
	 * Specifies State.
	 * 
	 * @param streamState conveys item stream state value
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.StreamState#OPEN})
	 * @param dataState conveys item data state value
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.DataState#SUSPECT})
	 * @param statusCode conveys specific item state code
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.StatusCode#NONE})
	 * @return reference to this object
	 */
	public RefreshMsg state(int streamState, int dataState, int statusCode);
	
	/**
	 * Specifies State.
	 * 
	 * @param streamState conveys item stream state value
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.StreamState#OPEN})
	 * @param dataState conveys item data state value
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.DataState#SUSPECT})
	 * @param statusCode conveys specific item state code
	 *        (default value is {@link com.thomsonreuters.ema.access.OmmState.StatusCode#NONE})
	 * @param statusText conveys item status explanation (default value is 'empty string')
	 * @return reference to this object
	 */
	public RefreshMsg state(int streamState, int dataState, int statusCode,	String statusText);

	/**
	 * Specifies SeqNum.
	 * 
	 * @param seqNum specifies sequence number
	 * @return reference to this object
	 */
	public RefreshMsg seqNum(long seqNum);

	/**
	 * Specifies PartNum.
	 * 
	 * @param partNum specifies part number
	 * @return reference to this object
	 */
	public RefreshMsg partNum(int partNum);

	/**
	 * Specifies ItemGroup.
	 * 
	 * @param itemGroup a ByteBuffer object with item group information
	 * @return reference to this object
	 */
	public RefreshMsg itemGroup(ByteBuffer itemGroup);

	/**
	 * Specifies PermissionData.
	 * 
	 * @param permissionData a ByteBuffer object with permission data information
	 * @return reference to this object
	 */
	public RefreshMsg permissionData(ByteBuffer permissionData);

	/**
	 * Specifies PublisherId.
	 * 
	 * @param userId specifies publisher's user id
	 * @param userAddress specifies publisher's user address
	 * @return reference to this object
	 */
	public RefreshMsg publisherId(long userId, long userAddress);

	/**
	 * Specifies Attrib.
	 * 
	 * @param data an object of ComplexType
	 * @return reference to this object
	 */
	public RefreshMsg attrib(ComplexType data);

	/**
	 * Specifies Payload.
	 * 
	 * @param data an object of ComplexType
	 * @return reference to this object
	 */
	public RefreshMsg payload(ComplexType data);

	/**
	 * Specifies ExtendedHeader.
	 * 
	 * @param buffer a ByteBuffer containing extendedHeader information
	 * @return reference to this object
	 */
	public RefreshMsg extendedHeader(ByteBuffer buffer);

	/**
	 * Specifies Solicited.
	 * 
	 * @param solicited true if this refresh is solicited; false otherwise
	 *            (default value is true)
	 * @return reference to this object
	 */
	public RefreshMsg solicited(boolean solicited);

	/**
	 * Specifies DoNotCache.
	 * 
	 * @param doNotCache true if this refresh must not be cached; false otherwise
	 *        (default value is false)
	 * @return reference to this object
	 */
	public RefreshMsg doNotCache(boolean doNotCache);

	/**
	 * Specifies ClearCache.
	 * 
	 * @param clearCache true if cache needs to be cleared; false otherwise
	 *            (default value is false)
	 * @return reference to this object
	 */
	public RefreshMsg clearCache(boolean clearCache);

	/**
	 * Specifies RefreshComplete.
	 * 
	 * @param complete true if this is the last part of multi part refresh
	 *            or single part refresh; false otherwise (default value is true)
	 * @return reference to this object
	 */
	public RefreshMsg complete(boolean complete);

	/**
	 * Specifies PrivateStream.
	 * 
	 * @param privateStream true if private stream; false otherwise
	 *        (default value is false)
	 * @return reference to this object
	 */
	public RefreshMsg privateStream(boolean privateStream);
}

File: Cpp-C/Ema/Src/Access/Impl/RefreshMsg.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2015-2026 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

#include "RefreshMsg.h"
#include "RefreshMsgImpl.h"

#include "OmmState.h"
#include "Utilities.h"
#include "GlobalPool.h"
#include "RdmUtilities.h"
#include "StaticDecoder.h"

using namespace refinitiv::ema::access;
using namespace refinitiv::ema::rdm;

extern const EmaString& getDTypeAsString( DataType::DataTypeEnum dType );

RefreshMsg::RefreshMsg() :
 Msg(),
 _toString()
{
	setImpl( g_pool.getRefreshMsgImplItem() );
}

RefreshMsg::RefreshMsg(UInt32 size) :
 RefreshMsg()
{
	impl()->Arena().reserve(size);
}

RefreshMsg::RefreshMsg(const RefreshMsg& other) :
 Msg(),
 _toString()
{
	setImpl( g_pool.getRefreshMsgImplItem() );
	*impl() = *other.impl();
}

RefreshMsg::RefreshMsg( RefreshMsg&& other ) noexcept :
 Msg(),
 _toString()
{
	setImpl( other.impl() );
	other.setImpl( nullptr );
}

RefreshMsg& RefreshMsg::operator=( const RefreshMsg& other )
{
	if (this == &other)
		return *this;

	*impl() = *other.impl();

	return *this;
}

RefreshMsg& RefreshMsg::operator=( RefreshMsg&& other ) noexcept
{
	if (this == &other)
		return *this;

	if ( impl() )
	{
		impl()->deallocateBuffers<RefreshMsgImpl>();
		g_pool.returnItem(impl());
	}

	setImpl( other.impl() );
	other.setImpl( nullptr );

	return *this;
}

RefreshMsg::~RefreshMsg()
{
	if ( impl() )
	{
		impl()->deallocateBuffers<RefreshMsgImpl>();
		g_pool.returnItem( impl() );

		setImpl( nullptr );
	}
}

RefreshMsg& RefreshMsg::clear()
{
	EMA_ASSERT( impl() != nullptr, "Private implementation can't be missing");

	impl()->clear();

	return *this;
}

Data::DataCode RefreshMsg::getCode() const
{
	return Data::NoCodeEnum;
}

DataType::DataTypeEnum RefreshMsg::getDataType() const
{
	return DataType::RefreshMsgEnum;
}

const EmaString& RefreshMsg::toString() const
{
	return toString( 0 );
}

const EmaString& RefreshMsg::toString( const refinitiv::ema::rdm::DataDictionary& dictionary ) const
{
	if (!dictionary.isEnumTypeDefLoaded() || !dictionary.isFieldDictionaryLoaded())
		return _toString.clear().append("\nDictionary is not loaded.\n");

	RefreshMsg refreshMsg;

	RsslBuffer& rsslBuffer = getEncoder().getEncodedBuffer();

	StaticDecoder::setRsslData( &refreshMsg, &rsslBuffer,
								RSSL_DT_MSG, RSSL_RWF_MAJOR_VERSION, RSSL_RWF_MINOR_VERSION,
								dictionary._pImpl->rsslDataDictionary() );

	_toString.clear().append(refreshMsg.toString());

	return _toString;
}

const EmaString& RefreshMsg::toString( UInt64 indent ) const
{
	_toString.clear();

	addIndent( _toString, indent++ ).append( "RefreshMsg" );
	addIndent( _toString, indent, true ).append( "streamId=\"" ).append( getStreamId() ).append( "\"" );
	addIndent( _toString, indent, true ).append( "domain=\"" ).append( rdmDomainToString( getDomainType() ) ).append( "\"" );

	if ( getSolicited() )
		addIndent( _toString, indent, true ).append( "Solicited" );

	if ( getComplete() )
		addIndent( _toString, indent, true ).append( "RefreshComplete" );

	if ( getPrivateStream() )
		addIndent( _toString, indent, true ).append( "PrivateStream" );

	if ( getDoNotCache() )
		addIndent( _toString, indent, true ).append( "DoNotCache" );

	if ( getClearCache() )
		addIndent( _toString, indent, true ).append( "ClearCache" );

	addIndent( _toString, indent, true ).append( "state=\"" ).append( getState().toString() ).append( "\"" );

	EmaString temp;
	hexToString( temp, getItemGroup() );
	addIndent( _toString, indent, true ).append( "itemGroup=\"" ).append( temp ).append( "\"" );

	if ( hasPermissionData() )
	{
		EmaString temp;
		hexToString( temp, getPermissionData() );
		addIndent( _toString, indent, true ).append( "permissionData=\"" ).append( temp ).append( "\"" );
	}

	if ( hasQos() )
		addIndent( _toString, indent, true ).append( "qos=\"" ).append( getQos().toString() ).append( "\"" );

	if ( hasPartNum() )
		addIndent( _toString, indent, true ).append( "partNum=\"" ).append( getPartNum() ).append( "\"" );

	if ( hasSeqNum() )
		addIndent( _toString, indent, true ).append( "seqNum=\"" ).append( getSeqNum() ).append( "\"" );

	if ( hasPublisherId() )
	{
		addIndent( _toString, indent, true ).append( "publisher user address=\"" ).append( getPublisherIdUserAddress() ).append( "\"" );
		addIndent( _toString, indent, true ).append( "publisher user id=\"" ).append( getPublisherIdUserId() ).append( "\"" );
	}

	indent--;
	if ( hasMsgKey() )
	{
		indent++;
		if ( hasName() )
			addIndent( _toString, indent, true ).append( "name=\"" ).append( getName() ).append( "\"" );

		if ( hasNameType() )
			addIndent( _toString, indent, true ).append( "nameType=\"" ).append( getNameType() ).append( "\"" );

		if ( hasServiceId() )
			addIndent( _toString, indent, true ).append( "serviceId=\"" ).append( getServiceId() ).append( "\"" );

		if ( hasServiceName() )
			addIndent( _toString, indent, true ).append( "serviceName=\"" ).append( getServiceName() ).append( "\"" );

		if ( hasFilter() )
			addIndent( _toString, indent, true ).append( "filter=\"" ).append( getFilter() ).append( "\"" );

		if ( hasId() )
			addIndent( _toString, indent, true ).append( "id=\"" ).append( getId() ).append( "\"" );

		indent--;

		if ( impl()->hasAttrib<const RefreshMsgImpl>() )
		{
			indent++;
			addIndent( _toString, indent, true ).append( "Attrib dataType=\"" ).append( getDTypeAsString( impl()->getAttribData().getDataType() ) ).append( "\"\n" );

			indent++;
			_toString.append( impl()->getAttribData().toString( indent ) );
			indent--;

			addIndent( _toString, indent, true ).append( "AttribEnd" );
			indent--;
		}
	}

	if ( hasExtendedHeader() )
	{
		indent++;
		addIndent( _toString, indent, true ).append( "ExtendedHeader\n" );

		indent++;

		addIndent( _toString, indent );
		hexToString( _toString, getExtendedHeader() );

		indent--;

		addIndent( _toString, indent, true ).append( "ExtendedHeaderEnd" );
		indent--;
	}

	if ( impl()->hasPayload() )
	{
		indent++;
		addIndent( _toString, indent, true ).append( "Payload dataType=\"" ).append( getDTypeAsString( impl()->getPayloadData().getDataType() ) ).append( "\"\n" );

		indent++;
		_toString.append( impl()->getPayloadData().toString( indent ) );
		indent--;

		addIndent( _toString, indent, true ).append( "PayloadEnd" );
		indent--;
	}

	addIndent( _toString, indent, true ).append( "RefreshMsgEnd\n" );

	return _toString;
}

const EmaBuffer& RefreshMsg::getAsHex() const
{
	return impl()->getAsHex();
}

bool RefreshMsg::hasQos() const
{
	return impl()->hasQos();
}

bool RefreshMsg::hasSeqNum() const
{
	return impl()->hasSeqNum();
}

bool RefreshMsg::hasPartNum() const
{
	return impl()->hasPartNum();
}

bool RefreshMsg::hasPermissionData() const
{
	return impl()->hasPermissionData();
}

bool RefreshMsg::hasPublisherId() const
{
	return impl()->hasPublisherId();
}

bool RefreshMsg::hasServiceName() const
{
	return impl()->hasServiceName();
}

const OmmState& RefreshMsg::getState() const
{
	return impl()->getState();
}

const OmmQos& RefreshMsg::getQos() const
{
	return impl()->getQos();
}

UInt32 RefreshMsg::getSeqNum() const
{
	return impl()->getSeqNum();
}

UInt16 RefreshMsg::getPartNum() const
{
	return impl()->getPartNum();
}

const EmaBuffer& RefreshMsg::getItemGroup() const
{
	return impl()->getItemGroup();
}

const EmaBuffer& RefreshMsg::getPermissionData() const
{
	return impl()->getPermissionData();
}

UInt32 RefreshMsg::getPublisherIdUserId() const
{
	return impl()->getPublisherIdUserId();
}

UInt32 RefreshMsg::getPublisherIdUserAddress() const
{
	return impl()->getPublisherIdUserAddress();
}

bool RefreshMsg::getDoNotCache() const
{
	return impl()->getDoNotCache();
}

bool RefreshMsg::getSolicited() const
{
	return impl()->getSolicited();
}

bool RefreshMsg::getComplete() const
{
	return impl()->getComplete();
}

bool RefreshMsg::getClearCache() const
{
	return impl()->getClearCache();
}

bool RefreshMsg::getPrivateStream() const
{
	return impl()->getPrivateStream();
}

const EmaString& RefreshMsg::getServiceName() const
{
	return impl()->getServiceName();
}

Decoder& RefreshMsg::getDecoder()
{
	if ( !impl() )
	{
		setImpl( g_pool.getRefreshMsgImplItem() );
	}

	return impl()->_decoder;
}

RefreshMsg& RefreshMsg::streamId( Int32 streamId )
{
	impl()->setStreamId( streamId );
	return *this;
}

RefreshMsg& RefreshMsg::domainType( UInt16 domainType )
{
	if ( domainType > 255 )
	{
		EmaString temp( "Passed in DomainType is out of range." );
		throwDtuException( domainType, temp );
		return *this;
	}

	impl()->setDomainType( (UInt8)domainType );
	return *this;
}

RefreshMsg& RefreshMsg::name( const EmaString& name )
{
	impl()->setName( name );
	return *this;
}

RefreshMsg& RefreshMsg::nameType( UInt8 nameType )
{
	impl()->setNameType( nameType );
	return *this;
}

RefreshMsg& RefreshMsg::serviceName( const EmaString& serviceName )
{
	impl()->copyServiceName( serviceName );
	return *this;
}

RefreshMsg& RefreshMsg::serviceId( UInt32 serviceId )
{
	impl()->setServiceId( serviceId );
	return *this;
}

RefreshMsg& RefreshMsg::id( Int32 id )
{
	impl()->setId( id );
	return *this;
}

RefreshMsg& RefreshMsg::filter( UInt32 filter )
{
	impl()->setFilter( filter );
	return *this;
}

RefreshMsg& RefreshMsg::qos( UInt32 timeliness, UInt32 rate )
{

	impl()->setQos( timeliness, rate );
	return *this;
}

RefreshMsg& RefreshMsg::state( OmmState::StreamState streamState, OmmState::DataState dataState,
						UInt8 statusCode, const EmaString& statusText )
{

	impl()->setState( streamState, dataState, statusCode, statusText );
	return *this;
}

RefreshMsg& RefreshMsg::seqNum( UInt32 seqNum )
{

	impl()->setSeqNum( seqNum );
	return *this;
}

RefreshMsg& RefreshMsg::partNum( UInt16 partNum )
{

	impl()->setPartNum( partNum );
	return *this;
}

RefreshMsg& RefreshMsg::itemGroup( const EmaBuffer& itemGroup )
{

	impl()->setItemGroup( itemGroup );
	return *this;
}

RefreshMsg& RefreshMsg::permissionData( const EmaBuffer& permissionData )
{
	impl()->setPermissionData( permissionData );
	return *this;
}

RefreshMsg& RefreshMsg::publisherId( UInt32 userId, UInt32 userAddress )
{
	impl()->setPublisherId( userId, userAddress );
	return *this;
}

RefreshMsg& RefreshMsg::attrib( const ComplexType& data )
{
	impl()->setAttrib( data );
	return *this;
}

RefreshMsg& RefreshMsg::payload( const ComplexType& data )
{
	impl()->setPayload( data );
	return *this;
}

RefreshMsg& RefreshMsg::extendedHeader( const EmaBuffer& buffer )
{
	impl()->setExtendedHeader( buffer );
	return *this;
}

RefreshMsg& RefreshMsg::solicited( bool solicited )
{
	impl()->setSolicited( solicited );
	return *this;
}

RefreshMsg& RefreshMsg::doNotCache( bool doNotCache )
{

	impl()->setDoNotCache( doNotCache );
	return *this;
}

RefreshMsg& RefreshMsg::clearCache( bool clearCache )
{

	impl()->setClearCache( clearCache );
	return *this;
}

RefreshMsg& RefreshMsg::complete( bool complete )
{
	impl()->setComplete( complete );
	return *this;
}

RefreshMsg& RefreshMsg::privateStream( bool privateStream )
{
	impl()->setPrivateStream( privateStream );
	return *this;
}

RefreshMsgImpl* RefreshMsg::impl()
{
	return static_cast<RefreshMsgImpl*>(_pImpl);
}

const RefreshMsgImpl* RefreshMsg::impl() const
{
	return static_cast<const RefreshMsgImpl*>(_pImpl);
}
